File: src/my_project/interviews/top_150_questions_round_22/ex_69_construct_tree_from_preorder_inorder_traversal.py
from typing import List, Union, Collection, Mapping, Optional
from abc import ABC, abstractmethod

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        # Create a hashmap for quick lookup of indices in inorder
        inorder_map = {val: idx for idx, val in enumerate(inorder)}
        self.preorder_idx = 0
        
        def build(left: int, right: int) -> Optional[TreeNode]:
            # Base case: no elements to construct the tree
            if left > right:
                return None
            
            # Pick the current root from preorder
            root_val = preorder[self.preorder_idx]
            root = TreeNode(root_val)
            
            # Move to the next element in preorder
            self.preorder_idx += 1
            
            # Find the index of root in inorder to split left and right subtrees
            inorder_idx = inorder_map[root_val]
            
            # Build left subtree (all elements before root in inorder)
            root.left = build(left, inorder_idx - 1)
            
            # Build right subtree (all elements after root in inorder)
            root.right = build(inorder_idx + 1, right)
            
            return root
        
        return build(0, len(inorder) - 1)

File: src/my_project/interviews_typescript/top_150_questions_round_1/ex_69_construct_tree_from_preorder_inorder_traversal.ts
import { TreeNode } from './TreeNode';         

function buildTree(preorder: number[], inorder: number[]): TreeNode | null {
    // Create a hashmap for quick lookup of indices in inorder
    const inorderMap = new Map<number, number>();
    inorder.forEach((val, idx) => inorderMap.set(val, idx));
    
    let preorderIdx = 0;
    
    function build(left: number, right: number): TreeNode | null {
        // Base case: no elements to construct the tree
        if (left > right) {
            return null;
        }
        
        // Pick the current root from preorder
        const rootVal = preorder[preorderIdx];
        const root = new TreeNode(rootVal);
        
        // Move to the next element in preorder
        preorderIdx++;
        
        // Find the index of root in inorder to split left and right subtrees
        const inorderIdx = inorderMap.get(rootVal)!;
        
        // Build left subtree (all elements before root in inorder)
        root.left = build(left, inorderIdx - 1);
        
        // Build right subtree (all elements after root in inorder)
        root.right = build(inorderIdx + 1, right);
        
        return root;
    }
    
    return build(0, inorder.length - 1);
}

